internal/prehook/prehook.go

```diff
@@ -0,0 +1,175 @@
+// Package prehook implements a connection prehook mechanism, to handle any
+// pre-negotiation required by a remote node before the B2F protocol can
+// commence (e.g. packet node traversal).
+package prehook
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/la5nta/pat/internal/debug"
+	"github.com/la5nta/pat/internal/directories"
+	"golang.org/x/sync/errgroup"
+)
+
+var ErrConnNotWrapped = errors.New("connection not wrapped for prehook")
+
+type Script struct {
+	File string
+	Args []string
+	Env  []string
+}
+
+// Execute executes the prehook script on a wrapped connection.
+//
+// ErrConnNotWrapped is returned if conn is not wrapped.
+func (s Script) Execute(ctx context.Context, conn net.Conn) error {
+	if conn, ok := conn.(*Conn); ok {
+		return conn.Execute(ctx, s)
+	}
+	return ErrConnNotWrapped
+}
+
+type Conn struct {
+	net.Conn
+	br *bufio.Reader
+}
+
+// Verify returns nil if the given script file is found and valid.
+func Verify(file string) error { _, err := lookPath(file); return err }
+
+func lookPath(file string) (string, error) {
+	// Look in our custom location first
```

Comment:
This works, but in an earlier comment you described trying to append `prehooks` to the process' `PATH`. If that worked the way I assumed, our `prehooks` dir would just be the first place `LookPath` would explore, and you wouldn't have to do two checks. I'm guessing that didn't work out?

Reply:
Yes, that would be much easier. I thought the issue was that Go's standard library didn't provide a portable way of modifying $PATH. The syntax differs between Windows and linux/darwin.

.. However, while writing this comment I discovered [os.PathListSeparator](https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/os/path_unix.go;l=11) 😄 I'll give it another try! Thanks 🙂